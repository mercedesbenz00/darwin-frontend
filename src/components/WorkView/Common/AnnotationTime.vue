<template>
  <div
    v-tooltip="'Total time the annotator has spent on this image'"
    class="annotation-time"
  >
    <img
      class="annotation-time__icon"
      src="/static/imgs/clock.svg"
    >
    <div class="annotation-time__clock">
      {{ time }}
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

import { formatDurationAsTimer } from '@/utils'

/**
 * Shows total annotation time on a task, as a HH:MM:SS time display with an icon in front
 */
@Component({ name: 'annotation-time' })
export default class AnnotationTime extends Vue {
  @Prop({ required: false, default: null, type: Number })
  timeInSeconds!: number | null

  get time (): string {
    const { timeInSeconds } = this
    return formatDurationAsTimer(timeInSeconds || 0)
  }
}
</script>

<!-- eslint-disable vue-scoped-css/enforce-style-type -->
<style lang="scss">
.annotation-time {
  @include row--center;

}

.annotation-time__icon {
  margin-right: 5px;
  padding-bottom: 3px;
}

.annotation-time__clock {
  @include typography(md-1, default, 500);
}
</style>
