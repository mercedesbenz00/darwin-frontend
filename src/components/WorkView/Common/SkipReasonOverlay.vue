<template>
  <div class="overlay">
    {{ text }}
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

/**
 * Used to style a div overlayed on top of workview center,
 * which renders the skipped/discarded/rejected/archived message.
 *
 * To maintain reusability between new and classic workflow, this accepts a text prop.
 *
 * TODO:
 *
 * Once classic worview is done, we should use selected stage instance
 * and dataset item from state, to infer what should be rendered.
 *
 * We should also pass styling to workflow.
 */
@Component({ name: 'skip-reason-overlay' })
export default class SkipReasonOverlay extends Vue {
  @Prop({ required: true, type: String })
  text!: string
}

</script>

<style lang="scss" scoped>
.overlay {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  padding: 10px 20px;
  border-radius: 5px;
  border: 2px solid $colorWhite;
  @include typography(xl, default, bold);
  background: rgba(255, 40, 117, 0.5);
  color: $colorWhite;
}
</style>
