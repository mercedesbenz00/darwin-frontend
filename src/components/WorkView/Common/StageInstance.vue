<template>
  <status-button
    class="instance"
    :class="{'instance--selected': isSelected}"
    :type="instance.type"
    inverted
    @click="$emit('click')"
  >
    <team-member-avatar
      v-if="assignee"
      class="instance__avatar"
      :member="assignee"
    />
  </status-button>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import { Getter, State } from 'vuex-class'

import TeamMemberAvatar from '@/components/Common/Avatar/V1/ResponsiveTeamMemberAvatar.vue'
import StatusButton from '@/components/WorkView/Common/StatusButton/V2/StatusButton.vue'
import { MembershipPayload, WorkflowStagePayload } from '@/store/types'

@Component({ name: 'stage-instance', components: { StatusButton, TeamMemberAvatar } })
export default class StageInstance extends Vue {
  @Prop({ required: true, type: Object as () => WorkflowStagePayload })
  instance!: WorkflowStagePayload

  @State(state => state.workview.selectedStageInstance)
  selectedInstance!: WorkflowStagePayload | null

  get isSelected (): boolean {
    const { instance, selectedInstance } = this
    return !!selectedInstance && selectedInstance.id === instance.id
  }

  @Getter('relevantTeamMemberships', { namespace: 'team' })
  memberships!: MembershipPayload[]

  get assignee (): MembershipPayload | null {
    const { instance, memberships } = this
    return memberships.find(m => m.user_id === instance.assignee_id) || null
  }
}
</script>
<style lang="scss" scoped>
.instance {
  border-width: 2px;
}

.instance.status-button,
.instance:hover.status-button {
  @include workflow-status-background-color;
  color: $colorWhite;
}

</style>
