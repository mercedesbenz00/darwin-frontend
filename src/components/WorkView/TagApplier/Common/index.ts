export { default as TagApplierHeader } from './TagApplierHeader/TagApplierHeader.vue'
export { default as TagApplierInput } from './TagApplierInput/TagApplierInput.vue'
export { default as TagApplierList } from './TagApplierList/TagApplierList.vue'
export { default as TagApplierCreate } from './TagApplierCreate/TagApplierCreate.vue'
