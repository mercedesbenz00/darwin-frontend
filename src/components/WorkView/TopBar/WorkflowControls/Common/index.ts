export { default as ActionHistory } from './ActionHistory/ActionHistory.vue'
export { default as ActionHistoryTrigger } from './ActionHistory/ActionHistoryTrigger.vue'
export { default as V2ActionHistoryTrigger } from './ActionHistory/V2/ActionHistoryTrigger.vue'
export { default as ContinueButton } from './ContinueButton/ContinueButton.vue'
export { default as DiscardButton } from './DiscardButton/DiscardButton.vue'
export { default as StageArchiveButton } from './StageArchiveButton/StageArchiveButton.vue'
export { default as StageRejectButton } from './StageRejectButton/StageRejectButton.vue'
export { default as StageSkipButton } from './StageSkipButton/V1/StageSkipButton.vue'
export { default as V2StageSkipButton } from './StageSkipButton/V2/StageSkipButton.vue'
