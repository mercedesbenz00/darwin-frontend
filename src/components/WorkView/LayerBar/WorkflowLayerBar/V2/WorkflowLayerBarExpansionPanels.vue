<template>
  <expansion-panels>
    <expansion-panel
      v-for="({ viewsStageAnnotations, view }, index) in viewsStageAnnotations"
      :id="view.id"
      :key="view.id"
      active-by-default
      @click="focusOn = view.id"
    >
      <template
        v-if="viewsList.length > 1"
        #header
      >
        <div
          class="header"
          :class="{'header--active': focusOn === view.id}"
        >
          <active-section-icon
            :view="view"
            :active="focusOn === view.id"
          />
          <span>{{ view.fileManager.filename }} ({{ view.fileManager.slotName }})</span>
        </div>
      </template>

      <template #content>
        <slot
          name="content"
          :views-stage-annotations="viewsStageAnnotations"
          :view="view"
          :index="index"
        />
      </template>
    </expansion-panel>
  </expansion-panels>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue'

import { ExpansionPanels, ExpansionPanel } from '@/components/Common/ExpansionPanels'
import ActiveSectionIcon from '@/components/WorkView/LayerBar/WorkflowLayerBar/ActiveSectionIcon.vue'
import { useActiveView, useEditorLayout } from '@/composables/useEditorV2'
import { View } from '@/engineV2/models'
import { StageAnnotation } from '@/store/modules/workview/types'

export default defineComponent({
  name: 'WorkflowLayerBarAccordion',
  components: {
    ExpansionPanels,
    ExpansionPanel,
    ActiveSectionIcon
  },
  props: {
    viewsStageAnnotations: {
      required: true,
      type: Array as () => { viewsStageAnnotations: StageAnnotation[], view: View }[]
    }
  },
  setup () {
    const activeView = useActiveView()
    const layout = useEditorLayout()

    const viewsList = computed(() => layout.value.viewsList)

    const focusOn = computed({
      get (): string {
        return activeView.value.id
      },
      set (id: string | null) {
        if (id === null) { return }

        const view = viewsList.value.find(view => view.id === id)

        if (!view) { return }

        layout.value.setActiveView(view)
      }
    })

    return {
      viewsList,
      focusOn
    }
  }
})
</script>

<style lang="scss" scoped>
.header {
  font-size: 13px;
  display: grid;
  grid-template-columns: auto 1fr;
  align-items: center;
  gap: calc($defaultSpace / 2);

  &.header--active {
    color: $colorDarkBlue;
  }
}
</style>
