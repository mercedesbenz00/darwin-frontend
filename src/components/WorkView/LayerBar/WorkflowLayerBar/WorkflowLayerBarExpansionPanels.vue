<template>
  <expansion-panels>
    <expansion-panel
      v-for="({ stageAnnotations, view }, index) in viewsStageAnnotations"
      :id="view.id"
      :key="view.id"
      active-by-default
      @click="focusOn = view.id"
    >
      <template
        v-if="editor.viewsList.length > 1"
        #header
      >
        <div
          class="header"
          :class="{'header--active': focusOn === view.id}"
        >
          <active-section-icon
            :view="view"
            :active="focusOn === view.id"
          />
          <span v-if="view.currentItem">{{ view.currentItem.filename }} {{ index }}</span>
        </div>
      </template>

      <template #content>
        <slot
          name="content"
          :stage-annotations="stageAnnotations"
          :view="view"
          :index="index"
        />
      </template>
    </expansion-panel>
  </expansion-panels>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

import { ExpansionPanels, ExpansionPanel } from '@/components/Common/ExpansionPanels'
import { Editor } from '@/engine/editor'
import { View } from '@/engine/models'
import { StageAnnotation } from '@/store/modules/workview/types'

import ActiveSectionIcon from './ActiveSectionIcon.vue'

@Component({
  name: 'workflow-layer-bar-expansion-panels',
  components: {
    ExpansionPanels,
    ExpansionPanel,
    ActiveSectionIcon
  }
})
export default class WorkflowLayerBarAccordion extends Vue {
  @Prop({ required: true, type: Object as () => Editor })
  editor!: Editor

  @Prop({ required: true, type: Array })
  viewsStageAnnotations!: { stageAnnotations: StageAnnotation[], view: View }[]

  get focusOn (): string {
    return this.editor.activeView.id
  }

  set focusOn (id: string | null) {
    if (id === null) { return }

    const view = this.editor.viewsList.find(view => view.id === id)

    if (!view) { return }

    this.editor.layout.setActiveView(view)
  }
}
</script>

<style lang="scss" scoped>
.header {
  font-size: 13px;
  display: grid;
  grid-template-columns: auto 1fr;
  align-items: center;
  gap: calc($defaultSpace / 2);

  &.header--active {
    color: $colorDarkBlue;
  }
}
</style>
