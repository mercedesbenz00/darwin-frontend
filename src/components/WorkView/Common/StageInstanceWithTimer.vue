<template>
  <div class="instance-wrapper">
    <stage-instance
      class="instance-wrapper__instance"
      :instance="instance"
    />
    <timer
      v-if="instance.completes_at && nextStatus"
      :completes-at="instance.completes_at"
      :class="`instance-wrapper__timer--${nextStatus}`"
      class="instance-wrapper__timer"
      :show-cancel="false"
    />
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import { State, Getter } from 'vuex-class'

import { Timer } from '@/components/WorkView/Common/Timer'
import {
  WorkflowStagePayload,
  WorkflowTemplatePayload,
  StageActionType,
  DatasetItemStatus,
  StageType
} from '@/store/types'
import { StageActionTypeToStatus } from '@/utils/StageActionTypeToStatus'

import StageInstance from './StageInstance.vue'
import { StageInstanceWithTimerProps } from './types'

@Component({ name: 'stage-instance-with-timer', components: { StageInstance, Timer } })
export default class StageInstanceWithTimer extends Vue implements StageInstanceWithTimerProps {
  @Prop({ required: true, type: Object as () => WorkflowStagePayload })
  instance!: WorkflowStagePayload

  get progressing (): boolean {
    const { instance } = this
    if (instance.metadata.review_status === 'rejected') { return false }
    return true
  }

  @State(state => state.workview.workflowTemplates)
  workflowTemplates!: WorkflowTemplatePayload[]

  get workflowTemplate (): WorkflowTemplatePayload | null {
    const { instance, workflowTemplates } = this
    const template = workflowTemplates.find(
      w => w.workflow_stage_templates.some(t => t.id === instance.workflow_stage_template_id)
    )

    return template || null
  }

  @Getter('stageCurrentAction', { namespace: 'workview' })
  getContinueAction!: (stage: WorkflowStagePayload) => StageActionType

  get nextStatus (): DatasetItemStatus | 'rejected' | StageType | null {
    const { instance } = this
    if (instance.type === StageType.Complete) { return null }
    const nextAction = this.getContinueAction(instance)
    return StageActionTypeToStatus[nextAction]
  }
}
</script>

<style lang="scss" scoped>
.instance-wrapper {
  position: relative;
  height: 100%;
  width: 100%;

  :deep(.instance) {
    height: 100%;
    width: 100%;
  }

  > * {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translateX(-50%) translateY(-50%);
  }

  &__timer {
    @include workflow-status-color;
  }
}
</style>
