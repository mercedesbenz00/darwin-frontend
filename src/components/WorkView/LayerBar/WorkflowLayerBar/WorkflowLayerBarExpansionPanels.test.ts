import { createLocalVue, shallowMount } from '@vue/test-utils'
import * as uuid from 'uuid'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'

import { ExpansionPanels } from '@/components/Common/ExpansionPanels'
import { Editor } from '@/engine/editor'
import { ItemManager } from '@/engine/managers'

import WorkflowLayerBarExpansionPanels from './WorkflowLayerBarExpansionPanels.vue'

const localVue = createLocalVue()

localVue.use(Vuex)

let store: ReturnType<typeof createTestStore>

let propsData: {
  editor: Editor,
  viewsStageAnnotations: Array<any>
}

const stubs = {
  ExpansionPanel: true,
  ExpansionPanels
}

jest.mock('uuid')
let counter = 0
jest.spyOn(uuid, 'v4').mockImplementation(() =>
  `fake-id-${counter++}`
)

beforeEach(() => {
  counter = 0
  store = createTestStore()

  const editor = new Editor(new ItemManager(store), store)
  editor.setupLayout({
    type: 'horizontal',
    views: [{
      item: null,
      framesGroup: [0]
    }, {
      item: null,
      framesGroup: [1]
    }]
  })

  propsData = {
    editor,
    viewsStageAnnotations: editor.viewsList.map(view => ({
      stageAnnotations: [],
      view
    }))
  }
})

it('matches snapshot', () => {
  const wrapper = shallowMount(
    WorkflowLayerBarExpansionPanels,
    { localVue, stubs, store, propsData }
  )
  expect(wrapper).toMatchSnapshot()
})
