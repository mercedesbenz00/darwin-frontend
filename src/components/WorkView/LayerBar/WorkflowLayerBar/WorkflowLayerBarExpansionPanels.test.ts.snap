// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<ul class="expansion-panels">
  <expansionpanel-stub id="fake-id-5" activebydefault="true"></expansionpanel-stub>
  <expansionpanel-stub id="fake-id-6" activebydefault="true"></expansionpanel-stub>
</ul>
`;
